nntrainer/layers/cl_layers/transpose_cl.cpp

```diff
@@ -0,0 +1,357 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2024 Niket Agarwal <niket.a@samsung.com>
+ *
+ * @file   transpose_cl.cpp
+ * @date   31 July 2024
+ * @brief  Implementation of transpose layer
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Niket Agarwal <niket.a@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+
+#include "transpose_cl.h"
+#include <iostream>
+#include <layer_context.h>
+#include <nntrainer_error.h>
+#include <nntrainer_log.h>
+#include <node_exporter.h>
+
+std::string transpose_cl_kernel_fp16_ =
+  R"(
+    #pragma OPENCL EXTENSION cl_khr_fp16 : enable
+    __kernel void transpose_cl_fp16(__global const half* in, 
+                               __global half* output,
+                               const int N, 
+                               const int C, 
+                               const int H, 
+                               const int W) {
+
+    int input_size = H * W;
+    int output_size = H * W;
+
+    // Calculate n, c, h, w from the global ids
+    int h = get_global_id(0);
+    int w = get_global_id(1);
+
+    if (h < H && w < W) {
+        for (int c = 0; c < C; ++c) {
+            for (int n = 0; n < N; ++n) {
+                // Calculate the input and output indices
+                int input_index = n * (C * input_size) + c * input_size + h * W + w;
+                int output_index = c * (N * output_size) + n * output_size + h * W + w;
```

Comment:
does the current kernel support both formats, channels last and channels first, or designed for channels first only?

Reply:
Only channels first. (nchw)